### PR TITLE
Avoid implicit float-to-int conversions

### PR DESCRIPTION
**Pull request by kronbichler**
This pull request allows me to finally compile ExaDG without warnings. The compiler complained about the following kind of double-to-float conversions (to give one representative example):
```
In file included from /home/kronbichler/Work/dg/exadg/applications/incompressible_navier_stokes/fda_benchmark/solver.cpp:26:
/home/kronbichler/Work/dg/exadg/applications/incompressible_navier_stokes/fda_benchmark/application.h:196:36: warning: implicit conversion turns floating-point number into integer: 'double' to 'unsigned int' [-Wfloat-conversion]
      n_points = 20 * (degree + 1) * std::pow(2.0, refine_space);
               ~ ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
What happens is that (at least for clang) `std::pow` takes a floating point argument and produces a floating point result, which is propagated through the multiplication and finally assigned to the integer `n_points`. That implicit conversion is what the compiler complains about. This PR addresses conversions by two approaches:
- Use deal.II's dedicated integer power mechanism `Utilities::pow` for integer powers, e.g. `(degree+1)^dim`, to avoid creating floating point numbers in the first place
- Where floating point numbers are necessary, e.g. for `round(log(n) / log(2))`, I added explicit casts.

### CODE REVIEW

File: include/exadg/postprocessor/kinetic_energy_spectrum.cpp [hunk 1]
```diff
@@ -204,12 +205,16 @@ class DealSpectrumWrapper
...
       timer.start("Permutation");
-      ArrayView<double> dst(
-        fftw.u_real,
-        s.dim * pow_(static_cast<types::global_dof_index>(s.cells * s.points_dst), s.dim) * 2);
+      ArrayView<double>       dst(fftw.u_real,
+                            s.dim *
+                              dealii::Utilities::pow(static_cast<types::global_dof_index>(
+                                                       s.cells * s.points_dst),
+                                                     s.dim) *
+                              2);
```
Comment (nfehn) [inline include/exadg/postprocessor/kinetic_energy_spectrum.cpp:213]: Is there a way to refactor this code so that the alignment is not that ugly? A first try would be to introduce an empty line between the variables `dst` and `src_` so that there is more space left for the `dst` variable. Or, why not remove the namespace dealii from this file to make the expressions shorter?

File: include/exadg/postprocessor/kinetic_energy_spectrum.cpp [hunk 2]
```diff
@@ -250,31 +255,20 @@ class DealSpectrumWrapper
...
-  pow_(const types::global_dof_index base, const types::global_dof_index exp)
-  {
-    types::global_dof_index result = 1.0;
-
-    for(types::global_dof_index i = 0; i < exp; i++)
-      result *= base;
-
-    return result;
-  };
-
```
Comment (nfehn) [inline include/exadg/postprocessor/kinetic_energy_spectrum.cpp:263]: This is no longer used?

File: include/exadg/postprocessor/spectral_analysis/bijection.h [hunk 3]
```diff
@@ -206,9 +207,8 @@ class Bijection
         int temp_counter = counter;
         for(int d = 0; d < s.dim; d++)
         {
-          int r = temp_counter % n;
+          Y[d] = temp_counter % n;
           temp_counter /= n;
-          Y[d] = r + 0.5;
```
Comment (nfehn) [inline include/exadg/postprocessor/spectral_analysis/bijection.h:211]: Can you briefly explain why this change is legal?

File: include/exadg/utilities/hypercube_resolution_parameters.h [hunk 4]
```diff
@@ -77,7 +77,7 @@ fill_resolutions_vector(
   int    refine_level = 0;
   double n_cells      = 1.0;
 
-  while(n_cells <= std::pow(2, dim) * n_cells_max)
+  while(n_cells <= std::pow(2.0, dim) * n_cells_max)
```
Comment (nfehn) [inline include/exadg/utilities/hypercube_resolution_parameters.h:80]: why didn't you switch to `int` for this file?

File: include/exadg/postprocessor/spectral_analysis/bijection.h (hunk 3 above)
Comment (kronbichler) [reply include/exadg/postprocessor/spectral_analysis/bijection.h:211]: I was wondering about this one, too. My reasoning is that `r` is a non-negative integer, and `Y` is an array of integers. So I cannot see that `(int)(r + 0.5)` gives anything else than `r`, so one can simply cast that back. I did not try to really understand the algorithm here. Do we have a good test for this?

Comment (nfehn) [reply include/exadg/postprocessor/spectral_analysis/bijection.h:211]: Let's ask @peterrum ?

File: include/exadg/postprocessor/kinetic_energy_spectrum.cpp (hunk 2 above)
Comment (kronbichler) [reply include/exadg/postprocessor/kinetic_energy_spectrum.cpp:263]: Indeed, this is the same as `Utilities::pow()` - I made sure that the argument in `pow` is `types::global_dof_index` to also work with large sizes.

File: include/exadg/postprocessor/kinetic_energy_spectrum.cpp (hunk 1 above)
Comment (kronbichler) [reply include/exadg/postprocessor/kinetic_energy_spectrum.cpp]: I introduced a separate variable `size`, see c6b0cb09

File: include/exadg/utilities/hypercube_resolution_parameters.h (hunk 4 above)
Comment (kronbichler) [reply include/exadg/utilities/hypercube_resolution_parameters.h]: Can you have a look at 3442131d?

File: include/exadg/utilities/hypercube_resolution_parameters.h [hunk 5]
```diff
@@ -71,13 +71,13 @@ fill_resolutions_vector(
 {
   unsigned int l = 0, n_subdivisions_1d = 1;
 
-  double n_cells_min = (double)n_dofs_min / dofs_per_element;
-  double n_cells_max = (double)n_dofs_max / dofs_per_element;
+  types::global_dof_index n_cells_min = n_dofs_min / dofs_per_element;
+  types::global_dof_index n_cells_max = (n_dofs_max + dofs_per_element - 1) / dofs_per_element;
```
Comment (nfehn) [inline include/exadg/utilities/hypercube_resolution_parameters.h:75]: I am currently not sure if this is correct. The user expects that `n_dofs_min <= n_dofs <= n_dofs_max` for all `n_dofs` realized in the throughput experiment. So we would have to ceil `n_cells_min` and floor `n_cells_max`?

Comment (kronbichler) [reply include/exadg/utilities/hypercube_resolution_parameters.h:75]: I agree, my formula did it the wrong way around (floor/ceil), including slightly too much.

File: include/exadg/postprocessor/spectral_analysis/bijection.h (hunk 3 above)
Comment (nfehn) [reply include/exadg/postprocessor/spectral_analysis/bijection.h:211]: @peterrum would you like to open an issue or pull request for that? I think we should keep this out of the present PR.

**Review by nfehn — APPROVED** [review]

**Review by peterrum — APPROVED** [review]
